iroh-gossip/src/proto/plumtree.rs

```diff
@@ -0,0 +1,462 @@
+    fn hash<H: std::hash::Hasher>(&self, state: &mut H) {
+        state.write(&self.0);
+    }
+}
+
+impl PartialEq for MessageId {
+    fn eq(&self, other: &Self) -> bool {
+        self.0 == other.0
+    }
+}
+
+impl fmt::Display for MessageId {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        let mut text = data_encoding::BASE32_NOPAD.encode(&self.0);
+        text.make_ascii_lowercase();
+        write!(f, "{}", text)
+    }
+}
+impl fmt::Debug for MessageId {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        let mut text = data_encoding::BASE32_NOPAD.encode(&self.0);
+        text.make_ascii_lowercase();
+        write!(f, "{}…{}", &text[..5], &text[(text.len() - 2)..])
+    }
+}
+
+#[derive(
+    From, Add, Sub, Serialize, Deserialize, Eq, PartialEq, PartialOrd, Ord, Clone, Copy, Debug, Hash,
+)]
+pub struct Round(u16);
+
+impl Round {
+    pub fn next(&self) -> Round {
+        Round(self.0 + 1)
+    }
+}
+
+#[derive(Debug, Serialize, Deserialize, Clone)]
+pub enum Message {
+    /// When receiving Gossip, emit as event and forward full message to eager peer
+    /// and (after a delay) message IDs to lazy peers.
+    Gossip(Gossip),
+    /// When receiving Prune, move the peer from the eager to the lazy set
+    Prune,
+    /// When receiving Graft, move the peer to the eager set and send the full content for the
+    /// included message ID.
+    Graft(Graft),
+    /// When receiving IHave, do nothing initially, and request the messages for the included
+    /// message IDs after some time if they aren't pushed eagerly to us.
+    IHave(Vec<IHave>),
+}
+
+#[derive(Serialize, Deserialize, Clone)]
+pub struct Gossip {
+    id: MessageId,
+    round: Round,
+    content: Bytes,
+}
+impl Gossip {
+    pub fn next_round(self) -> Gossip {
+        Gossip {
+            id: self.id,
+            content: self.content,
+            round: self.round.next(),
+        }
+    }
+}
+impl fmt::Debug for Gossip {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("Gossip")
+            .field("id", &self.id)
+            .field("round", &self.round)
+            .field("content", &format!("<{}>", self.content.len()))
+            .finish()
+    }
+}
+
+#[derive(Debug, Serialize, Deserialize, Clone)]
+pub struct IHave {
+    id: MessageId,
+    round: Round,
+}
+
+#[derive(Debug, Serialize, Deserialize, Clone)]
+pub struct Graft {
+    id: Option<MessageId>,
+    round: Round,
+}
+
+#[derive(Clone, Debug)]
+pub struct Config {
+    pub graft_timeout_1: Duration,
+    pub graft_timeout_2: Duration,
+    pub dispatch_timeout: Duration,
+    pub optimization_threshold: Round,
+}
+impl Default for Config {
+    fn default() -> Self {
+        Self {
+            graft_timeout_1: Duration::from_millis(80),
+            graft_timeout_2: Duration::from_secs(40),
+            dispatch_timeout: Duration::from_millis(50),
+            optimization_threshold: Round(7),
```

Comment:
would be good to document/describe these defaults

Reply:
Yup, good point. I pushed two commits that document the default values for plumtree and hyparview. Note: They will have to be validated in experiments or practice; their source is either the defaults from the paper; or wild guesses.
Also: They will depend on which network size and sender/receiver distrubtion we are optimizing for. Some implementations derive them from an `expected_network_size` parameter.